File: typescript/libs/workspace/adapters/json/system/src/lib/runtime/DatabaseRecord.ts
import { Class, RelationType, RoleType } from '@allors/workspace/meta/system';
import { DatabaseRecord as SystemDatabaseRecord, has, IRange } from '@allors/workspace/adapters/system';
import { SyncResponseObject, SyncResponseRole } from '@allors/protocol/json/system';
import { Database } from './Database';
import { ResponseContext } from './Security/ResponseContext';
import { unitFromJson } from '../json/fromJson';

export class DatabaseRecord extends SystemDatabaseRecord {
  accessControlIds: IRange;
  deniedPermissionIds: IRange;

  private _roleByRelationType?: Map<RelationType, unknown>;
  private syncResponseRoles?: SyncResponseRole[];

  constructor(public readonly database: Database, cls: Class, id: number, public version: number) {
    super(cls, id, version);
  }

  static fromResponse(database: Database, ctx: ResponseContext, syncResponseObject: SyncResponseObject): DatabaseRecord {
    const obj = new DatabaseRecord(database, database.configuration.metaPopulation.metaObjectByTag.get(syncResponseObject.t) as Class, syncResponseObject.i, syncResponseObject.v);
    obj.syncResponseRoles = syncResponseObject.r;
    obj.accessControlIds = ctx.checkForMissingAccessControls(syncResponseObject.a);
    obj.deniedPermissionIds = ctx.checkForMissingPermissions(syncResponseObject.d);
    return obj;
  }

  get roleByRelationType(): Map<RelationType, unknown> {
    if (this.syncResponseRoles != null) {
      const metaPopulation = this.database.configuration.metaPopulation;
      this._roleByRelationType = new Map(
        this.syncResponseRoles.map((v) => {
          const relationType = metaPopulation.metaObjectByTag.get(v.t) as RelationType;
          const roleType = relationType.roleType;
          const objectType = roleType.objectType;
          let role: unknown;

          if (objectType.isUnit) {
            role = unitFromJson(objectType.tag, v.v);
          } else {
            if (roleType.isOne) {
              role = v.o;
            } else {
              role = v.c;
            }
          }

          return [relationType, role];
        })
      );

      delete this.syncResponseRoles;
    }

    return this._roleByRelationType;
  }

  getRole(roleType: RoleType): unknown {
    return this.roleByRelationType?.get(roleType.relationType);
  }

  isPermitted(permission: number): boolean {
    if (this.accessControlIds == null) {
      return false;
    }

    if (has(this.deniedPermissionIds, permission)) {
      return false;
    }

    if (this.accessControlIds == null) {
      return false;
    }

    return this.accessControlIds.some((v) => has(this.database.accessControlById.get(v).permissionIds, permission));
  }
}

File: dotnet/System/workspace/csharp/adapters/allors.workspace.adapters.remote/database/DatabaseRecord.cs
// <copyright file="RemoteDatabaseObject.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System.Collections.Generic;
    using System.Linq;
    using Allors.Protocol.Json.Api.Sync;
    using Meta;
    using Ranges;

    internal class DatabaseRecord : Adapters.DatabaseRecord
    {
        private readonly DatabaseConnection database;

        private Dictionary<IRelationType, object> roleByRelationType;
        private SyncResponseRole[] syncResponseRoles;

        internal DatabaseRecord(DatabaseConnection database, IClass @class, long id, long version) : base(@class, id, version) => this.database = database;

        internal static DatabaseRecord FromResponse(DatabaseConnection database, ResponseContext ctx, SyncResponseObject syncResponseObject) =>
            new DatabaseRecord(database, (IClass)database.Configuration.MetaPopulation.FindByTag(syncResponseObject.t), syncResponseObject.i, syncResponseObject.v)
            {
                syncResponseRoles = syncResponseObject.r,
                AccessControlIds = database.Ranges.Load(ctx.CheckForMissingAccessControls(syncResponseObject.a)),
                DeniedPermissions = database.Ranges.Load(ctx.CheckForMissingPermissions(syncResponseObject.d))
            };

        internal IRange AccessControlIds { get; private set; }

        internal IRange DeniedPermissions { get; private set; }

        private Dictionary<IRelationType, object> RoleByRelationType
        {
            get
            {
                if (this.syncResponseRoles != null)
                {
                    var ranges = this.database.Ranges;
                    var metaPopulation = this.database.Configuration.MetaPopulation;
                    this.roleByRelationType = this.syncResponseRoles.ToDictionary(
                        v => (IRelationType)metaPopulation.FindByTag(v.t),
                        v =>
                        {
                            var roleType = ((IRelationType)metaPopulation.FindByTag(v.t)).RoleType;
                            var objectType = roleType.ObjectType;

                            if (objectType.IsUnit)
                            {
                                return this.database.UnitConvert.FromJson(objectType.Tag, v.v);
                            }

                            if (roleType.IsOne)
                            {
                                return v.o;
                            }

                            return ranges.Load(v.c);
                        });

                    this.syncResponseRoles = null;
                }

                return this.roleByRelationType;
            }
        }

        public override object GetRole(IRoleType roleType)
        {
            object @object = null;
            this.RoleByRelationType?.TryGetValue(roleType.RelationType, out @object);
            return @object;
        }

        public override bool IsPermitted(long permission)
        {
            if (this.AccessControlIds == null)
            {
                return false;
            }

            return !this.DeniedPermissions.Contains(permission) && this.AccessControlIds.Any(v => this.database.AccessControlById[v].PermissionIds.Contains(permission));
        }
    }
}
